### PR TITLE
Fix scalar indexing when using ForwardDIff

### PR DESCRIPTION
**Pull request by MilkshakeForReal**
Note that the issue always exists and is not recently introduced. 
```julia
using Lux, Random, NNlib, Zygote, CUDA, ComponentArrays
CUDA.allowscalar(false)
using ForwardDiff
model = Chain(Dense(2 => 4))
rng = Random.default_rng()
x = randn(rng, 2, 4) |> gpu
ps, st = Lux.setup(rng, model)
ps = ps |> ComponentArray |> gpu
st = st |> gpu
julia> ForwardDiff.gradient(ps -> sum(first(model(x, ps,st))), ps)
ERROR: Scalar indexing is disallowed.
Invocation of setindex! resulted in scalar indexing of a GPU array.
This is typically caused by calling an iterating implementation of a method.
Such implementations *do not* execute on the GPU, but very slowly on the CPU,
and therefore are only permitted from the REPL for prototyping purposes.
If you did intend to index this array, annotate the caller with @allowscalar.
Stacktrace:
  [1] error(s::String)
    @ Base .\error.jl:33
  [2] assertscalar(op::String)
    @ GPUArraysCore C:\Users\Luffy\.julia\packages\GPUArraysCore\rSIl2\src\GPUArraysCore.jl:78
  [3] setindex!
    @ C:\Users\Luffy\.julia\packages\GPUArrays\gok9K\src\host\indexing.jl:17 [inlined]
  [4] copyto_unaliased!(deststyle::IndexLinear, dest::CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, srcstyle::IndexLinear, src::ForwardDiff.Partials{12, Float32})
    @ Base .\abstractarray.jl:1018
  [5] copyto!
    @ .\abstractarray.jl:998 [inlined]
  [6] copyto!
    @ C:\Users\Luffy\.julia\packages\ComponentArrays\IQNQW\src\similar_convert_copy.jl:36 [inlined]
  [7] extract_gradient!(#unused#::Type{ForwardDiff.Tag{var"#25#26", Float32}}, result::ComponentVector{Float32, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, Tuple{Axis{(weight = ViewAxis(1:8, ShapedAxis((4, 2), NamedTuple())), bias = ViewAxis(9:12, ShapedAxis((4, 1), NamedTuple())))}}}, dual::ForwardDiff.Dual{ForwardDiff.Tag{var"#25#26", Float32}, Float32, 12})
    @ ForwardDiff C:\Users\Luffy\.julia\packages\ForwardDiff\wAaVJ\src\gradient.jl:81
  [8] vector_mode_gradient(f::var"#25#26", x::ComponentVector{Float32, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, Tuple{Axis{(weight = ViewAxis(1:8, ShapedAxis((4, 2), NamedTuple())), bias = ViewAxis(9:12, ShapedAxis((4, 1), NamedTuple())))}}}, cfg::ForwardDiff.GradientConfig{ForwardDiff.Tag{var"#25#26", Float32}, Float32, 12, ComponentVector{ForwardDiff.Dual{ForwardDiff.Tag{var"#25#26", Float32}, Float32, 12}, CuArray{ForwardDiff.Dual{ForwardDiff.Tag{var"#25#26", Float32}, Float32, 12}, 1, CUDA.Mem.DeviceBuffer}, Tuple{Axis{(weight = ViewAxis(1:8, ShapedAxis((4, 2), NamedTuple())), bias = ViewAxis(9:12, ShapedAxis((4, 1), NamedTuple())))}}}})
    @ ForwardDiff C:\Users\Luffy\.julia\packages\ForwardDiff\wAaVJ\src\gradient.jl:109
  [9] gradient(f::Function, x::ComponentVector{Float32, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, Tuple{Axis{(weight = ViewAxis(1:8, ShapedAxis((4, 2), NamedTuple())), bias = ViewAxis(9:12, ShapedAxis((4, 1), NamedTuple())))}}}, cfg::ForwardDiff.GradientConfig{ForwardDiff.Tag{var"#25#26", Float32}, Float32, 12, ComponentVector{ForwardDiff.Dual{ForwardDiff.Tag{var"#25#26", Float32}, Float32, 12}, CuArray{ForwardDiff.Dual{ForwardDiff.Tag{var"#25#26", Float32}, Float32, 12}, 1, CUDA.Mem.DeviceBuffer}, Tuple{Axis{(weight = ViewAxis(1:8, ShapedAxis((4, 2), NamedTuple())), bias = ViewAxis(9:12, ShapedAxis((4, 1), NamedTuple())))}}}}, ::Val{true})
    @ ForwardDiff C:\Users\Luffy\.julia\packages\ForwardDiff\wAaVJ\src\gradient.jl:19
 [10] gradient(f::Function, x::ComponentVector{Float32, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, Tuple{Axis{(weight = ViewAxis(1:8, ShapedAxis((4, 2), NamedTuple())), bias = ViewAxis(9:12, ShapedAxis((4, 1), NamedTuple())))}}}, cfg::ForwardDiff.GradientConfig{ForwardDiff.Tag{var"#25#26", Float32}, Float32, 12, ComponentVector{ForwardDiff.Dual{ForwardDiff.Tag{var"#25#26", Float32}, Float32, 12}, CuArray{ForwardDiff.Dual{ForwardDiff.Tag{var"#25#26", Float32}, Float32, 12}, 1, CUDA.Mem.DeviceBuffer}, Tuple{Axis{(weight = ViewAxis(1:8, ShapedAxis((4, 2), NamedTuple())), bias = ViewAxis(9:12, ShapedAxis((4, 1), NamedTuple())))}}}}) (repeats 2 times)
    @ ForwardDiff C:\Users\Luffy\.julia\packages\ForwardDiff\wAaVJ\src\gradient.jl:17
 [11] top-level scope
    @ REPL[66]:1
 [12] top-level scope
    @ C:\Users\Luffy\.julia\packages\CUDA\DfvRa\src\initialization.jl:52

### CODE REVIEW

**Comment by MilkshakeForReal** [comment]
The culprit is `ForwardDiff.Partials` wrapping a `Tuple`, which makes CUDA unhappy
```julia
julia> using CUDA

julia> CUDA.allowscalar(false)

julia> a = CUDA.rand(3)
3-element CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}:
 0.64399
 0.62868804
 0.14182384

julia> copyto!(a,(4,5,6))
ERROR: Scalar indexing is disallowed.
Invocation of setindex! resulted in scalar indexing of a GPU array.
This is typically caused by calling an iterating implementation of a method.
Such implementations *do not* execute on the GPU, but very slowly on the CPU,
and therefore are only permitted from the REPL for prototyping purposes.
If you did intend to index this array, annotate the caller with @allowscalar.
Stacktrace:
 [1] error(s::String)
   @ Base .\error.jl:33
 [2] assertscalar(op::String)
   @ GPUArraysCore C:\Users\Luffy\.julia\packages\GPUArraysCore\rSIl2\src\GPUArraysCore.jl:78
 [3] setindex!
   @ C:\Users\Luffy\.julia\packages\GPUArrays\gok9K\src\host\indexing.jl:17 [inlined]
 [4] setindex!
   @ C:\Users\Luffy\.julia\packages\GPUArrays\gok9K\src\host\indexing.jl:24 [inlined]
 [5] copyto!(dest::CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, src::Tuple{Int64, Int64, Int64})
   @ Base .\abstractarray.jl:893
 [6] top-level scope
   @ REPL[30]:1
 [7] top-level scope
   @ C:\Users\Luffy\.julia\packages\CUDA\DfvRa\src\initialization.jl:52

```

**Comment by jonniedie** [comment]
Hey, thanks for the PR. I'm not actually sure if this fixes anything, though, because the only non-whitespace change adds a method for something that already worked (see comment).

**Review by jonniedie — CHANGES_REQUESTED** [review]

File: src/compat/forwarddiff.jl [hunk 1]
```diff
@@ -0,0 +1,4 @@
+function Base.copyto!(dest::ComponentArray, src::ForwardDiff.Partials{N,V}) where {N,V}
+    copyto!(getdata(dest), collect(src))
+    return dest
+end
```
[the remark this answers is not in the document]
Comment (MilkshakeForReal) [reply src/compat/forwarddiff.jl:4]: Yeah, because scalar indexing is fine on the CPU. Did you read my description?

**Comment by MilkshakeForReal** [comment]
I'm closing this since I will move it to Lux

**Comment by MilkshakeForReal** [comment]
reopen this since gup testing is finally here.

File: src/compat/forwarddiff.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (MilkshakeForReal) [reply src/compat/forwarddiff.jl:4]: Only dispatch on GPUComponentArray now

**Comment by jonniedie** [comment]
I wonder if we can just make another general method for `copy to!` like [this](https://github.com/jonniedie/ComponentArrays.jl/blob/f3a1b6d9e6d1549f7513579632d66597067eb93b/src/similar_convert_copy.jl#L35) that `collect`s the second argument in the case where it’s not an `AbstractArray`. That way we wouldn’t need a new `Requires` entry.

**Comment by MilkshakeForReal** [comment]
But we don't always need to collect. `Tuple` is fine on CPU for example and collecting makes it slow.

**Comment by MilkshakeForReal** [comment]
And `ForwardDiff.Partials` is an `AbstractArray`, so...
```
julia> ForwardDiff.Partials <: AbstractArray
true
```

**Comment by jonniedie** [comment]
Yeah, I’d ideally like to not collect stack allocated things. It was just a thought.

But this is weird. We’re already doing `getdata` on the ComponentArray for the default implementation of `copyto!`. Maybe this is an issue of `ForwardDiff.Partials` not doing the right thing?

**Comment by MilkshakeForReal** [comment]
Yeah this is weird. `Forward.gradient` don't support native CUDA arrays. 
```julia
julia> ForwardDiff.gradient(sum, cu[1,2])
ERROR: Scalar indexing is disallowed.
Invocation of setindex! resulted in scalar indexing of a GPU array.
This is typically caused by calling an iterating implementation of a method.
Such implementations *do not* execute on the GPU, but very slowly on the CPU,
and therefore are only permitted from the REPL for prototyping purposes.
If you did intend to index this array, annotate the caller with @allowscalar.
Stacktrace:
```
but `ForwardDiff.jacobian` does 
```julia
julia> ForwardDiff.jacobian(identity, cu[1,2])
2×2 CuArray{Int64, 2, CUDA.Mem.DeviceBuffer}:
 1  0
 0  1
```

So we should follow this and only support `ForwardDiff.jacobian` on GPU
```julia
julia> ps = ComponentArray(cu[1f0,2f0,3f0,4f0], Axis(a=1:2, b=3:4));

julia> ForwardDiff.jacobian(identity, ps)
ERROR: This object is not a GPU array
Stacktrace:
  [1] error(s::String)
    @ Base .\error.jl:33
  [2] backend(#unused#::Type)
    @ GPUArrays C:\Users\Luffy\.julia\packages\GPUArrays\Hyss4\src\device\execution.jl:15
  [3] backend(x::Matrix{Float32})
```
This is easy since `jacobian` should return a `Matrix`

File: src/compat/gpuarrays.jl [hunk 2]
```diff
@@ -1,4 +1,4 @@
-const GPUComponentArray = ComponentArray{T,N,<:GPUArrays.AbstractGPUArray,Ax} where {T,N,Ax}
+const GPUComponentArray = ComponentArray{T,N,<:GPUArrays.AbstractGPUArray,Ax} where {T,N,Ax<:Tuple{Vararg{AbstractAxis}}}
```
Comment (jonniedie) [inline src/compat/gpuarrays.jl:1]: Why?

**Review by jonniedie — COMMENTED** [review]
Hm. I think I want to dig into this a little more because It seems like the other broadcasting changes should have fixed this. I’d like to not have to add another Requires entry and if this is something that can be solved upstream with a change to ForwardDiff, it would benefit other array wrapper packages.

**Comment by MilkshakeForReal** [comment]
Closing this since we should have a better solution!

**Pull request closed without merging**